source/features/conversation-navigation.tsx

```diff
@@ -0,0 +1,164 @@
+import React from 'dom-chef';
+import cache from 'webext-storage-cache';
+import select from 'select-dom';
+import * as pageDetect from 'github-url-detection';
+import ChevronLeftIcon from 'octicon/chevron-left.svg';
+import ChevronRightIcon from 'octicon/chevron-right.svg';
+
+import features from '.';
+import * as api from '../github-helpers/api';
+import SearchQuery from '../github-helpers/search-query';
+import looseParseInt from '../helpers/loose-parse-int';
+import {getRepoURL, getConversationNumber} from '../github-helpers';
+
+function getButton(direction: string): HTMLAnchorElement {
+	return (
+		<a
+			aria-disabled="true"
+			className="btn btn-lg BtnGroup-item"
+			aria-label={`Navigate to ${direction} Conversation`}
+			data-hotkey={direction.charAt(0)}
+		>
+			{direction === 'next' ? <ChevronRightIcon/> : <ChevronLeftIcon/>}
+		</a>
+	) as unknown as HTMLAnchorElement;
+}
+
+async function init() {
+	const previousButton = getButton('previous');
+	const nextButton = getButton('next');
+
+	select('.gh-header-meta')!.append(
+		<div className="BtnGroup ml-2">
+			{previousButton}
+			{nextButton}
+		</div>
+	);
+
+	const conversationNumber = getConversationNumber();
+	const {list, query, page} = await getConversationList(conversationNumber);
+	const conversation = list.find(item => item.number === conversationNumber);
+
+	setButtonHref(
+		previousButton,
+		query,
+		page,
+		list.find(item => item.cursor === conversation!.cursor - 1)
+	);
+
+	setButtonHref(
+		nextButton,
+		query,
+		page,
+		list.find(item => item.cursor === conversation!.cursor + 1)
+	);
+}
+
+function setButtonHref(button: HTMLAnchorElement, query: string, page: number, conversation?: Conversation) {
+	if (conversation) {
+		const url = new URL(conversation.url);
+		url.searchParams.set('q', query);
+		url.searchParams.set('page', Number(page));
+		button.removeAttribute('aria-disabled');
+		button.href = url.href;
+	}
+}
+
+type Conversation = {
+	cursor: number;
+	number: number;
+	url: string;
+};
+
+const fetchConversationList = cache.function(async (query: string, page: number): Promise<Conversation[]> => {
+	// When fetching a page of conversations we want to also fetch last item from previous page, and first item from next page.
+	// This is needed for cases when conversation is first/last in list, and we need to know if items exist on previous/next page.
+	// To take into consideration also, that when we are on first item on first page, we don't have a previous item, and we don't want to fetch using a negative cursor.
+	const {search} = await api.v4(`
+		search(
+			type: ISSUE,
+			first: ${page === 1 ? ITEMS_PER_PAGE + 1 : ITEMS_PER_PAGE + 2},
+			after: "${btoa(`cursor:${page === 1 ? 0 : ((page - 1) * ITEMS_PER_PAGE) - 1}`)}",
+			query: "${query}"
+		) {
+			edges {
+				cursor
+				node {
+					... on Issue { number, url }
+					... on PullRequest { number, url }
+				}
+			}
+		}
+	`);
+
+	return search.edges.map((edge: any) => ({
```

Comment:
Would it be even easier if, instead of returning an array of objects, you return a array-like object?

```js
{
  3: [number, url],
  4: [number, url],
  5: [number, url],
  6: [number, url], etc
}
```

Basically

```js
const list = {};
for (const edge of search.edges) {
  list[looseParseInt(atob(edge.cursor))] = edge.node
}
return list;
```

This would save you from having to use `.find` over and over, because you can just always use `list[cursor]`

Reply:
I can do this, but then searching for current conversation looks ugly. Something like

```typescript
const conversation = Object.entries(list).find(([key, value]) => value[0] === conversationNumber)![1]
```